File: design-system/components/forms/Input.js
/**
 * Input Component
 *
 * A reusable input component with multiple variants and states.
 *
 * Usage:
 * ```
 * // Basic input
 * Input.text({ name: 'username', placeholder: 'Enter username' });
 *
 * // Input with label and icon
 * Input.text({
 *   name: 'email',
 *   label: 'Email Address',
 *   iconLeft: '<i data-lucide="mail" class="h-4 w-4 text-gray-400"></i>',
 *   placeholder: 'user@example.com'
 * });
 *
 * // Input with validation state
 * Input.text({
 *   name: 'password',
 *   type: 'password',
 *   label: 'Password',
 *   validationState: 'error',
 *   validationMessage: 'Password must be at least 8 characters'
 * });
 * ```
 */

class Input {
  constructor() {
    this.baseInputClasses =
      "appearance-none block w-full border border-gray-300 rounded-lg shadow-sm placeholder-gray-400 focus:outline-none focus:ring-cyan-500 focus:border-cyan-500 sm:text-sm";
    this.baseLabelClasses = "block text-sm font-medium text-gray-700 mb-1";
    this.baseIconClasses =
      "absolute inset-y-0 flex items-center pointer-events-none";
  }

  /**
   * Renders an input with the specified options
   * @param {Object} options - Input options
   * @param {string} options.type - Input type (text, password, email, etc.)
   * @param {string} options.name - Input name
   * @param {string} options.id - Input id (defaults to name if not provided)
   * @param {string} options.label - Input label
   * @param {string} options.placeholder - Input placeholder
   * @param {string} options.value - Input value
   * @param {boolean} options.required - Whether the input is required
   * @param {string} options.validationState - Validation state (success, error)
   * @param {string} options.validationMessage - Validation message
   * @param {string} options.iconLeft - Icon to show at the left
   * @param {string} options.iconRight - Icon to show at the right
   * @param {boolean} options.disabled - Whether the input is disabled
   * @param {string} options.className - Additional classes to apply
   * @returns {string} HTML markup for the input
   */
  render({
    type = "text",
    name,
    id = "",
    label = "",
    placeholder = "",
    value = "",
    required = false,
    validationState = "",
    validationMessage = "",
    iconLeft = "",
    iconRight = "",
    disabled = false,
    className = "",
  }) {
    if (!name) {
      throw new Error("Input name is required");
    }

    // Set id to name if not provided
    id = id || name;

    // Determine validation-specific classes
    let validationClasses = "";
    let validationColor = "";

    switch (validationState) {
      case "success":
        validationClasses =
          "border-green-500 focus:border-green-500 focus:ring-green-500";
        validationColor = "text-green-600";
        break;
      case "error":
        validationClasses =
          "border-red-500 focus:border-red-500 focus:ring-red-500";
        validationColor = "text-red-600";
        break;
      default:
        validationClasses = "";
        validationColor = "";
    }

    // Determine padding based on icons
    let paddingClasses = "";
    if (iconLeft && iconRight) {
      paddingClasses = "pl-10 pr-10";
    } else if (iconLeft) {
      paddingClasses = "pl-10 pr-3";
    } else if (iconRight) {
      paddingClasses = "pl-3 pr-10";
    } else {
      paddingClasses = "px-3";
    }

    // Combine all input classes
    const inputClasses = [
      this.baseInputClasses,
      paddingClasses,
      validationClasses,
      disabled ? "bg-gray-100 cursor-not-allowed" : "",
      className,
    ]
      .filter(Boolean)
      .join(" ");

    // Create HTML for the label if provided
    const labelHtml = label
      ? `<label for="${id}" class="${this.baseLabelClasses}">${label}${
          required ? ' <span class="text-red-500">*</span>' : ""
        }</label>`
      : "";

    // Create HTML for icons if provided
    const leftIconHtml = iconLeft
      ? `<div class="${this.baseIconClasses} left-0 pl-3">${iconLeft}</div>`
      : "";

    const rightIconHtml = iconRight
      ? `<div class="${this.baseIconClasses} right-0 pr-3">${iconRight}</div>`
      : "";

    // Create HTML for validation message if provided
    const validationHtml = validationMessage
      ? `<p class="mt-1 text-xs ${validationColor}">${validationMessage}</p>`
      : "";

    // Build the complete input HTML
    return `
      <div class="w-full">
        ${labelHtml}
        <div class="relative">
          ${leftIconHtml}
          <input
            type="${type}"
            id="${id}"
            name="${name}"
            ${placeholder ? `placeholder="${placeholder}"` : ""}
            ${value ? `value="${value}"` : ""}
            ${required ? "required" : ""}
            ${disabled ? "disabled" : ""}
            class="${inputClasses}"
          />
          ${rightIconHtml}
        </div>
        ${validationHtml}
      </div>
    `;
  }

  /**
   * Creates a text input
   */
  text(options = {}) {
    return this.render({ type: "text", ...options });
  }

  /**
   * Creates a password input
   */
  password(options = {}) {
    // Default icon for password fields if none provided
    const defaultRightIcon =
      '<button type="button" class="text-gray-400 hover:text-gray-500 focus:outline-none">' +
      '<i data-lucide="eye" class="h-4 w-4"></i>' +
      "</button>";

    return this.render({
      type: "password",
      iconRight: options.iconRight || defaultRightIcon,
      ...options,
    });
  }

  /**
   * Creates an email input
   */
  email(options = {}) {
    return this.render({ type: "email", ...options });
  }

  /**
   * Creates a number input
   */
  number(options = {}) {
    return this.render({ type: "number", ...options });
  }

  /**
   * Creates a search input
   */
  search(options = {}) {
    const defaultLeftIcon =
      '<i data-lucide="search" class="h-4 w-4 text-gray-400"></i>';

    return this.render({
      type: "search",
      iconLeft: options.iconLeft || defaultLeftIcon,
      ...options,
    });
  }
}

// Export the Input class
export default new Input();

File: app/components/input.py
"""
Input component that adheres to the shastho design system.
"""

def text(name, label=None, placeholder="", value="", required=False, error=None, helper_text=None,
         disabled=False, additional_classes=""):
    """
    Generate HTML for a text input field.

    Args:
        name (str): Input name/id
        label (str): Input label text
        placeholder (str): Placeholder text
        value (str): Current value
        required (bool): Whether the field is required
        error (str): Error message if validation failed
        helper_text (str): Additional helper text
        disabled (bool): Whether the input is disabled
        additional_classes (str): Additional CSS classes

    Returns:
        str: HTML for the text input
    """
    # Generate a unique ID if name contains spaces or special characters
    input_id = name.replace(" ", "_").lower()

    # Base container
    html = f'<div class="mb-4">'

    # Add label if provided
    if label:
        html += f'<label for="{input_id}" class="block text-sm font-medium text-gray-700 mb-1">{label}'
        if required:
            html += ' <span class="text-error">*</span>'
        html += '</label>'

    # Input classes
    input_classes = "w-full rounded-lg border px-3 py-2 text-gray-700 focus:outline-none focus:ring-2 focus:ring-primary focus:border-transparent"

    # Add error classes if there's an error
    if error:
        input_classes += " border-error focus:ring-error"
    else:
        input_classes += " border-gray-300"

    # Add disabled classes
    if disabled:
        input_classes += " bg-gray-100 cursor-not-allowed"

    # Add any additional classes
    if additional_classes:
        input_classes += f" {additional_classes}"

    # Build the input element
    required_attr = 'required' if required else ''
    disabled_attr = 'disabled' if disabled else ''

    html += f'<input type="text" id="{input_id}" name="{name}" placeholder="{placeholder}" '
    html += f'value="{value}" class="{input_classes}" {required_attr} {disabled_attr}>'

    # Add helper text or error message
    if error:
        html += f'<p class="mt-1 text-sm text-error">{error}</p>'
    elif helper_text:
        html += f'<p class="mt-1 text-sm text-gray-500">{helper_text}</p>'

    html += '</div>'

    return html

def password(name, label=None, placeholder="", required=False, error=None, helper_text=None,
             disabled=False, additional_classes=""):
    """
    Generate HTML for a password input field.

    Args:
        name (str): Input name/id
        label (str): Input label text
        placeholder (str): Placeholder text
        required (bool): Whether the field is required
        error (str): Error message if validation failed
        helper_text (str): Additional helper text
        disabled (bool): Whether the input is disabled
        additional_classes (str): Additional CSS classes

    Returns:
        str: HTML for the password input
    """
    # Generate a unique ID from the name if needed
    input_id = name.lower().replace(' ', '_')

    # Start building the HTML
    html = '<div class="mb-4">'

    # Add label if provided
    if label:
        html += f'<label for="{input_id}" class="block text-sm font-medium text-gray-700 mb-1">{label}'
        if required:
            html += ' <span class="text-error">*</span>'
        html += '</label>'

    # Base input classes
    input_classes = "w-full px-3 py-2 border rounded-lg text-gray-700 focus:outline-none focus:ring-2 focus:ring-primary/20 focus:border-primary"

    # Add error or disabled styles
    if error:
        input_classes += " border-error focus:border-error focus:ring-error/20"
    elif disabled:
        input_classes += " bg-gray-100 border-gray-300 text-gray-500 cursor-not-allowed"
    else:
        input_classes += " border-gray-300"

    # Add user's additional classes
    if additional_classes:
        input_classes += f" {additional_classes}"

    # Required attribute
    required_attr = 'required' if required else ''

    # Disabled attribute
    disabled_attr = 'disabled' if disabled else ''

    # Build the input tag
    html += f'<input type="password" id="{input_id}" name="{name}" placeholder="{placeholder}" '
    html += f'class="{input_classes}" {required_attr} {disabled_attr}>'

    # Add error message if provided
    if error:
        html += f'<p class="mt-1 text-sm text-error">{error}</p>'

    # Add helper text if provided
    if helper_text and not error:
        html += f'<p class="mt-1 text-sm text-gray-500">{helper_text}</p>'

    html += '</div>'

    return html

def select(name, options, label=None, selected="", required=False, error=None, helper_text=None,
           disabled=False, additional_classes=""):
    """
    Generate HTML for a select dropdown.

    Args:
        name (str): Select name/id
        options (list): List of options as dicts with 'value' and 'text' keys
        label (str): Select label text
        selected (str): Currently selected value
        required (bool): Whether the field is required
        error (str): Error message if validation failed
        helper_text (str): Additional helper text
        disabled (bool): Whether the select is disabled
        additional_classes (str): Additional CSS classes

    Returns:
        str: HTML for the select dropdown
    """
    # Generate a unique ID from the name if needed
    select_id = name.lower().replace(' ', '_')

    # Start building the HTML
    html = '<div class="mb-4">'

    # Add label if provided
    if label:
        html += f'<label for="{select_id}" class="block text-sm font-medium text-gray-700 mb-1">{label}'
        if required:
            html += ' <span class="text-error">*</span>'
        html += '</label>'

    # Base select classes
    select_classes = "w-full px-3 py-2 border rounded-lg text-gray-700 focus:outline-none focus:ring-2 focus:ring-primary/20 focus:border-primary"

    # Add error or disabled styles
    if error:
        select_classes += " border-error focus:border-error focus:ring-error/20"
    elif disabled:
        select_classes += " bg-gray-100 border-gray-300 text-gray-500 cursor-not-allowed"
    else:
        select_classes += " border-gray-300"

    # Add user's additional classes
    if additional_classes:
        select_classes += f" {additional_classes}"

    # Required attribute
    required_attr = 'required' if required else ''

    # Disabled attribute
    disabled_attr = 'disabled' if disabled else ''

    # Build the select tag and options
    html += f'<select id="{select_id}" name="{name}" class="{select_classes}" {required_attr} {disabled_attr}>'

    for option in options:
        is_selected = 'selected' if option['value'] == selected else ''
        html += f'<option value="{option["value"]}" {is_selected}>{option["text"]}</option>'

    html += '</select>'

    # Add error message if provided
    if error:
        html += f'<p class="mt-1 text-sm text-error">{error}</p>'

    # Add helper text if provided
    if helper_text and not error:
        html += f'<p class="mt-1 text-sm text-gray-500">{helper_text}</p>'

    html += '</div>'

    return html

def textarea(name, label=None, placeholder="", value="", rows=4, required=False,
             error=None, helper_text=None, disabled=False, additional_classes=""):
    """
    Generate HTML for a textarea field.

    Args:
        name (str): Input name/id
        label (str): Input label text
        placeholder (str): Placeholder text
        value (str): Current value
        rows (int): Number of rows to display
        required (bool): Whether the field is required
        error (str): Error message if validation failed
        helper_text (str): Additional helper text
        disabled (bool): Whether the textarea is disabled
        additional_classes (str): Additional CSS classes

    Returns:
        str: HTML for the textarea
    """
    # Generate a unique ID if name contains spaces or special characters
    input_id = name.replace(" ", "_").lower()

    # Base container
    html = f'<div class="mb-4">'

    # Add label if provided
    if label:
        html += f'<label for="{input_id}" class="block text-sm font-medium text-gray-700 mb-1">{label}'
        if required:
            html += ' <span class="text-error">*</span>'
        html += '</label>'

    # Textarea classes
    textarea_classes = "w-full rounded-lg border px-3 py-2 text-gray-700 focus:outline-none focus:ring-2 focus:ring-primary focus:border-transparent"

    # Add error classes if there's an error
    if error:
        textarea_classes += " border-error focus:ring-error"
    else:
        textarea_classes += " border-gray-300"

    # Add disabled classes
    if disabled:
        textarea_classes += " bg-gray-100 cursor-not-allowed"

    # Add any additional classes
    if additional_classes:
        textarea_classes += f" {additional_classes}"

    # Build the textarea element
    required_attr = 'required' if required else ''
    disabled_attr = 'disabled' if disabled else ''

    html += f'<textarea id="{input_id}" name="{name}" placeholder="{placeholder}" '
    html += f'rows="{rows}" class="{textarea_classes}" {required_attr} {disabled_attr}>{value}</textarea>'

    # Add helper text or error message
    if error:
        html += f'<p class="mt-1 text-sm text-error">{error}</p>'
    elif helper_text:
        html += f'<p class="mt-1 text-sm text-gray-500">{helper_text}</p>'

    html += '</div>'

    return html